cfg/libcerror.cfg

```diff
@@ -0,0 +1,128 @@
+<?xml version="1.0"?>
+<!-- Official repository: https://github.com/libyal/libcerror -->
+<!-- Typically included by: #include <libcerror.h> -->
+<def format="2">
+  <!-- int libcerror_error_initialize( libcerror_error_t **error,
+                                       int error_domain,
+                                       int error_code ); -->
+  <function name="libcerror_error_initialize">
+    <noreturn>false</noreturn>
+    <returnValue type="int"/>
+    <arg nr="1">
+      <not-null/>
+      <not-uninit/>
+    </arg>
+    <arg nr="2">
+      <not-uninit/>
+    </arg>
+    <arg nr="3">
+      <not-uninit/>
+    </arg>
+  </function>
+  <!-- void libcerror_error_set( libcerror_error_t **error,
+                                 int error_domain,
+                                 int error_code,
+                                 const char *format_string,
+                                 ... ); -->
+  <function name="libcerror_error_set">
+    <noreturn>false</noreturn>
+    <returnValue type="void"/>
+    <arg nr="1">
+      <not-null/>
+      <not-uninit/>
+    </arg>
+    <arg nr="2">
+      <not-uninit/>
+    </arg>
+    <arg nr="3">
+      <not-uninit/>
+    </arg>
+    <formatstr/>
+    <arg nr="4">
+      <not-null/>
+      <not-uninit/>
+      <formatstr/>
+      <strz/>
+    </arg>
+  </function>
+  <!-- void libcerror_error_free( libcerror_error_t **error ); -->
+  <function name="libcerror_error_free">
+    <noreturn>false</noreturn>
+    <returnValue type="void"/>
+    <arg nr="1">
+      <not-uninit/>
+    </arg>
+  </function>
+  <!-- int libcerror_error_resize( libcerror_internal_error_t *internal_error ); -->
+  <function name="libcerror_error_resize">
+    <noreturn>false</noreturn>
+    <returnValue type="int"/>
```

Comment:
Would you like to add the specification “`<use-retval/>`” also for this function description?

Reply:
No

Reply:
I would appreciate a bit more source code analysis and development attention also for complete error detection for the safe usage of this software library.

Reply:
yes .. good idea. I would also like to have such a tool. However I want to have a "no false positives" tool also. I *really* do not want to change the cppcheck philosophy.

feel free to create a new tool with a different philosophy. You can fork Cppcheck if you want and get a quick start.

if you create that tool I think it would be less popular than cppcheck.. most people want to have "no false positives". but there will definitely be users that are happy by your philosophy also.

Reply:
>However I want to have a "no false positives" tool also.

How much would the indication of a memory reallocation failure fit to this desire?


>feel free to create a new tool with a different philosophy.

Other development tools are already available which offer additional source code analysis capabilities.


>if you create that tool I think it would be less popular than cppcheck.

I am more curious on how the existing system infrastructure can be extended.

Reply:
> How much would the indication of a memory reallocation failure fit to this desire?

we have our philosophy. I do not want to destroy the usefulness of cppcheck.

> Other development tools are already available which offer additional source code analysis capabilities.

Exactly. I think that clang or clang-tidy might match your philosophy better. I am no expert, but as far as I know they more or less write a warning unless the code is safe. You are supposed to make the code explicitly safe.

> I am more curious on how the existing system infrastructure can be extended.

this has already been answered by several people.

Reply:
>I do not want to destroy the usefulness of cppcheck.

I would find it nice if this tool can be improved further at some places.


>I am no expert, …

I imagine that you achieved a remarkable software expertise through the years.


>You are supposed to make the code explicitly safe.

I hope that more adjustments can be achieved in such a design direction.

Reply:
> I hope that more adjustments can be achieved in such a design direction.

With additional heuristics that means all false positives are avoided I think we can discuss it.

Otherwise, make these suggestions in another tool instead.

Reply:
>With additional heuristics that means all false positives are avoided I think we can discuss it.

Will it become easier (or safer) to [detect memory reallocation failures](https://github.com/libyal/libcerror/blob/3367ce13d37cd1dd8ffac8f48cd9a4afa0866a85/libcerror/libcerror_error.c#L122 "Implementation of libcerror_error_resize()")?

Reply:
Why don't you take a look at flawfinder instead? I think that has a philosophy that you will be happy with.

Reply:
> Will it become easier (or safer) to [detect memory reallocation failures](https://github.com/libyal/libcerror/blob/3367ce13d37cd1dd8ffac8f48cd9a4afa0866a85/libcerror/libcerror_error.c#L122)?

I think minor improvements with better heuristics can be achieved.

We rather report 10 warnings and all are real bugs. Than 100 warnings and 90 are real bugs.

In flawfinder I think you will be a hero if you can change the database so more warnings are reported and those have a 90% signal to noise ratio.

Reply:
This analysis tool contains also specific software limitations so that I find Cppcheck's approaches more promising for desired extensions so far.

Reply:
what limitations?

Reply:
the fact is that if there are 90 bugs.. and you want to detect all those.. then you have to use another tool.

Cppcheck will never detect all the possible bugs.

Reply:
Information for [further software development considerations](https://dwheeler.com/flawfinder/#how_work "How does Flawfinder Work?"):
* “Flawfinder works by using a built-in database of C/C++ functions with well-known problems, …”
* “Nevertheless, flawfinder is fundamentally a naive program; it doesn’t even know about the data types of function parameters, and it certainly doesn’t do control flow or data flow analysis …”
* Programming language “Python”

-----

Cppcheck can increase also its detection capabilities considerably.

Reply:
those limitations does not sound like any problem if you want to warn whenever the result of `libcerror_error_resize` is not used.

Reply:
The safe function description and corresponding agreement on API properties is a general development (or documentation) challenge, isn't it?

Reply:
but your suggestions does not correspond to cppcheck philosophy. it seems to correspond to the flawfinder philosophy.

Reply:
I find your change blockage questionable.
I assume that more emotional factors influence such a feedback here.

Reply:
I truly and genuinely believe your `<use-retval/>` suggestions would fit much better in the flawfinder database.

don't get discouraged that it is "naive". Cppcheck does not have as robust AST as Clang but imho we do pretty well with what we have.

Reply:
Can the clarification of relevant API properties become easier if the function “libcerror_error_resize” will be compared to [the standard function “realloc”](https://en.cppreference.com/w/cpp/memory/c/realloc "API properties for std::realloc()")?

-----

>we do pretty well with what we have.

* This can be on the first glance.
* Our discussion seems to point recurring open issues out from other areas.